package com.aerofs.proto;

//removed reflection. this makes obfuscation possible
option optimize_for = LITE_RUNTIME;

message PBTPHeader {

    enum Type {
        DATAGRAM = 0;
        STREAM = 1;

        // Messages to maintain link liveness
        HEARTBEAT_CALL = 2;
        HEARTBEAT_REPLY = 3;
    }

    required Type type = 1;

    // only for stream messages
    optional PBStream stream = 2;

    // only in maxcast payload
    optional uint32 mcast_id = 3;

    // only for HEARTBEAT messages
    optional PBHeartbeat heartbeat = 4;
}

message PBStream {
    enum Type {
        PAYLOAD = 0;
        BEGIN_STREAM = 1;
        TX_ABORT_STREAM = 2;
        RX_ABORT_STREAM = 3;
        PAUSE_STREAM = 4;
        RESUME_STREAM = 5;
    }

    enum InvalidationReason {
        UPDATE_IN_PROGRESS = 0;
        ENDED = 1;
        STREAM_NOT_FOUND = 2;
        INTERNAL_ERROR = 3;
        CHOKE_ERROR = 5;
        OUT_OF_ORDER = 6;       // a chunk is out of order
        STORE_NOT_FOUND = 7;
    }

    required Type type = 1;
    required uint32 stream_id = 2;
    optional uint32 seq_num = 3;

    optional InvalidationReason reason = 4;
}

message PBHeartbeat {
    required int32 heartbeat_id = 1;
    required uint64 sent_time = 2;
}
